### DISCUSSION TITLE
Share Your Thoughts on Kloppy!

### DISCUSSION THREAD

**Discussion by koenvo** [PySport/kloppy#276] [Announcements]
# Share Your Thoughts on Kloppy!

Hello Kloppy Community! 👋

We have an upcoming contributors meeting on January 5th, 2024, focusing on the future of Kloppy. We understand that not everyone can join us in person, so we're reaching out to gather your valuable insights and suggestions.

### We'd Love to Hear Your Ideas and Feedback on:

- **Feature Requests**: What new features would you like to see in Kloppy?
- **Improvements**: Any areas where Kloppy could be improved or enhanced?
- **Use Cases**: How are you using Kloppy, and what can we do to support your work better?
- **Community and Collaboration**: Ideas for fostering a stronger Kloppy community and collaboration?

### How to Participate:

- **Reply to this Discussion**: Share your thoughts, ideas, and feedback directly in this thread.
- **Create a New Discussion**: If you have a detailed proposal or topic, feel free to start a new discussion and link it here.

Your input is invaluable to us, and we're eager to incorporate the community's perspective into our roadmap. Let's work together to make Kloppy even better!

Thanks for being a part of our community!

Best,  
Kloppy Development Team

**Comment by theanatron** [comment]
With kloppy, I was able to visualize SkillCorner tracking data and use tracking data in raw form much much faster. I believe that kloppy's attempt to standardize the preprocessing portion with in a workflow (for event, tracking, etc) is important because it will hopefully to get more people with a more "football" background to use insights/data to better their performance.

Not everyone has a computer science degree and kloppy helps bridge the gap between those highly technical with CS/DS skills and sporting analytics.

Going forward it would be beneficial to include other sports, like basketball and tennis since those two sports are heavy within the analytics space. Limb Tracking data from basket would be something great to include since it will most likely be used more in football going forward.

Example :
https://github.com/theanatron/skillcorner/blob/main/frames82-88.gif

Thanks for your hard work! We appreciate what you all do :) 

Ana De Souza

**Comment by tomboomstra** [comment]
Probably not relevant for the small user, but we have largely moved to spark especially for tracking data, which means we can now handle entire seasons of tracking data in the cloud (including streaming). I wrote pyspark code to transform tracab data to spark dfs, perhaps this would be something to add to kloppy.

**Comment by footballdaniel** [comment]
Big fan of Kloppy. Tbh, I learned quite a bit about package development / domain models just by inspecting kloppy.

**Feature request 1**:
Implement not just the deserializers, but also serializers for the different formats so that datasets can be transformed. For instance Skillcorner -> kloppy -> Sportec would be something great to have (and the other way around)

Motivation: Many projects that I am involved with are not agnostic of the provider. I have put forward to use kloppy in many cases as an intermediary format, but in some cases, I got some pushback against using the domain model of kloppy because the projects are in a different codebase (e.g., c#). It would be great if I was able to transform datasets through kloppy to another provider.


**Feature request 2**:
One of the recurring challenges is event/tracking synchronization across providers. For example Opta events and skillcorner tracking data. If kloppy could implement / have interop with an existing synchronization algorithm, that would save a lot of work, since this synchronization between event and tracking data is performed at many organizations currently.

**Comment by black6197** [comment]
Today, when I used the kloppy package, I found that several Premier League games didn't seem to be used. Before, they were all available, namely matchid=2500033 and 2500039. And there are matchid=2500021 teamid for 1628 games does not seem to work, of course, not only these games, after the run prompt： player_id=str(player["player"]["wyId"]),

TypeError: 'NoneType' object is not subscriptable；

Changing the computer and the network and uninstalling the kolppy still cannot solve this problem. At present, you can only load the local json files, but it takes longer to load the files locally. Please the author can solve the problem, I do not know and do not understand is my own operation problem or what problem, before all can run smoothly, but today suddenly appeared such a problem, the following is my source code:
from kloppy import wyscout
import pandas as pd
dataset = wyscout.load_open_data(
    match_id=2500033,
    # Optional arguments
    event_types=["pass"],#, "shot"],
    coordinates="wyscout"
)
data = dataset.to_df()
The rest of the game ids will work just fine.
Looking forward to the author's early reply, thank you very much, and very much for providing the public data, thank you very much!!


今天我在用kloppy包的时候发现有几场英超比赛似乎用不了，之前都是可以用的，分别是matchid=2500033和2500039。并且还有matchid=2500021的teamid为1628的比赛似乎用不了，当然不止这些比赛，运行后提示player_id=str(player["player"]["wyId"]),

TypeError: 'NoneType' object is not subscriptable
；
换了电脑以及网络以及卸载kolppy都仍是无法解决该问题，目前只能采用加载本地json文件是可以使用的，但是本地加载文件的话运行就需要更长时间了。麻烦作者能够解决一下该问题，我不清楚也不明白是我自己操作问题还是什么问题，之前都是可以顺利运行的，但是今天突然就出现这样的问题了，下面是我源代码：
from kloppy import wyscout
import pandas as pd
dataset = wyscout.load_open_data(
    match_id=2500033,
    # Optional arguments
    event_types=["pass"],#, "shot"],
    coordinates="wyscout"
)
data = dataset.to_df()

其余比赛id都是可以正常运行的。
期待作者早日回复，非常感谢，也非常感谢作者提供的公开数据，非常感谢！！

**Reply by black6197** [reply]
I know!!!! I have modified the code in the deserializer_v2.py to include the source code:
def _parse_team(raw_events, wyId: str, ground: Ground) -> Team:
    team = Team(
        team_id=wyId,
        name=raw_events["teams"][wyId]["team"]["officialName"],
        ground=ground,
    )
    team.players = [
        Player(
            player_id=str(player["player"]["wyId"]),
            team=team,
            jersey_no=None,
            first_name=player["player"]["firstName"],
            last_name=player["player"]["lastName"],
        )
        for player in raw_events["players"][wyId]
    ]
    return team
Modified to:
def _parse_team(raw_events, wyId: str, ground: Ground) -> Team:
    team = Team(
        team_id=wyId,
        name=raw_events["teams"][wyId]["team"]["officialName"],
        ground=ground,
    )
    team.players = []
    for player_info in raw_events["players"].get(wyId, []):
        try:
            player = Player(
                player_id=str(player_info["player"]["wyId"]),
                team=team,
                jersey_no=None,
                first_name=player_info["player"]["firstName"],
                last_name=player_info["player"]["lastName"],
            )
            team.players.append(player)
        except (KeyError, TypeError) as e:
            print(f"Skipping player due to missing information: {e}")
            continue
    return team
In this way, you can skip when the player value is missing (I do not need the player information). In this way, the modification can be done.
Hope the author attaches importance to this problem!
我知道了！！！
我在deserializer_v2.py中修改了代码，将源代码：
def _parse_team(raw_events, wyId: str, ground: Ground) -> Team:
    team = Team(
        team_id=wyId,
        name=raw_events["teams"][wyId]["team"]["officialName"],
        ground=ground,
    )
    team.players = [
        Player(
            player_id=str(player["player"]["wyId"]),
            team=team,
            jersey_no=None,
            first_name=player["player"]["firstName"],
            last_name=player["player"]["lastName"],
        )
        for player in raw_events["players"][wyId]
    ]
    return team
修改成了：
def _parse_team(raw_events, wyId: str, ground: Ground) -> Team:
    team = Team(
        team_id=wyId,
        name=raw_events["teams"][wyId]["team"]["officialName"],
        ground=ground,
    )
    team.players = []
    for player_info in raw_events["players"].get(wyId, []):
        try:
            player = Player(
                player_id=str(player_info["player"]["wyId"]),
                team=team,
                jersey_no=None,
                first_name=player_info["player"]["firstName"],
                last_name=player_info["player"]["lastName"],
            )
            team.players.append(player)
        except (KeyError, TypeError) as e:
            print(f"Skipping player due to missing information: {e}")
            continue
    return team

这样，当遇到球员值缺失的时候可以跳过（本身我研究的内容并不需要球员的信息）。这样进行修改后就能够进行操作了。
希望作者重视这个问题！

**Reply by black6197** [reply]
Examined the next individual matches, but this may affect the overall study of other data。
检查了下个别比赛有几个球员信息的缺失，但是这可能会影响整体其他数据的研究。

**Comment by AleksandrKosolapov** [comment]
Hi there. What's the best way to get a dataframe with all Champions League matches from the last 5 years? I need the most detailed minute-by-minute statistics possible, along with average bookmaker odds for any available markets.
